packages/indexer/src/plugins/kv.ts

```diff
@@ -1,16 +1,108 @@
+import assert from "node:assert";
+import type { Cursor, DataFinality } from "@apibara/protocol";
+import { type Database, type ISqlite, open } from "sqlite";
 import { useIndexerContext } from "../context";
 import { defineIndexerPlugin } from "../plugins";
+import { deserialize, serialize } from "../vcr";
 
-/** This plugin is a placeholder for a future key-value store plugin. */
-export function kv<TFilter, TBlock, TRet>() {
+type SqliteArgs = ISqlite.Config;
+
+export function kv<TFilter, TBlock, TRet>(args: SqliteArgs) {
   return defineIndexerPlugin<TFilter, TBlock, TRet>((indexer) => {
+    let db: Database;
+
     indexer.hooks.hook("run:before", async () => {
+      db = await open(args);
+      await db.exec(`
+        CREATE TABLE IF NOT EXISTS kvs (
+          from_block INTEGER NOT NULL,
+          to_block INTEGER,
+          k TEXT NOT NULL,
+          v BLOB NOT NULL,
+          PRIMARY KEY (from_block, k)
+        );
+      `);
+    });
+
+    indexer.hooks.hook("handler:before", async ({ finality, endCursor }) => {
+      const ctx = useIndexerContext();
+
+      assert(endCursor, new Error("endCursor cannot be undefined"));
+
+      ctx.kv = new KVStore(db, finality, endCursor);
+
+      await db.exec("BEGIN TRANSACTION");
+    });
+
+    indexer.hooks.hook("handler:after", async () => {
+      await db.exec("COMMIT TRANSACTION");
+
+      const ctx = useIndexerContext();
+
+      ctx.kv = null;
+    });
+
+    indexer.hooks.hook("handler:exception", async () => {
+      await db.exec("COMMIT TRANSACTION");
+
       const ctx = useIndexerContext();
-      ctx.kv = {};
+
+      ctx.kv = null;
     });
 
     indexer.hooks.hook("run:after", async () => {
       console.log("kv: ", useIndexerContext().kv);
     });
   });
 }
+
+export class KVStore {
+  constructor(
+    private _db: Database,
+    private _finality: DataFinality,
+    private _endCursor: Cursor,
+  ) {}
+
+  async get<T extends Record<string, unknown>>(key: string): Promise<T> {
```

Comment:
I think here `T` can be anything that is json encodeable (e.g. I often store just a number or a string).